### ISSUE TITLE
adler 32 rollsum, the s1 is set to 1 instead of zero

### ISSUE THREAD

**Issue by guotie** [librsync/librsync#42]
https://www.ietf.org/rfc/rfc1950.txt:

   8.2. The Adler-32 algorithm

      The Adler-32 algorithm is much faster than the CRC32 algorithm yet
      still provides an extremely low probability of undetected errors.

      The modulo on unsigned long accumulators can be delayed for 5552
      bytes, so the modulo operation time is negligible.  If the bytes
      are a, b, c, the second sum is 3a + 2b + c + 3, and so is position
      and order sensitive, unlike the first sum, which is just a
      checksum.  That 65521 is prime is important to avoid a possible
      large class of two-byte errors that leave the check unchanged.
      (The Fletcher checksum uses 255, which is not prime and which also
      makes the Fletcher check insensitive to single byte changes 0 <->
      255.)

      The sum s1 is initialized to 1 instead of zero to make the length
      of the sequence part of s2, so that the length does not have to be
      checked separately. (Any sequence of zeroes has a Fletcher
      checksum of zero.)

**Comment by sourcefrog** [comment]
To be clear, you're suggesting we should do a new format using Adler-32 as the weak sum?

**Comment by dbaarda** [comment]
I discussed this back in 2002;

https://lists.samba.org/archive/rsync/2002-November/004176.html

A quick search for stuff about this and the wikipedia page says Fletcher32 normally included a modulus 2^16 - 1 and a "block-size" of 16. There is also some docs that suggest 1's compliment addition works better. There is also a bunch of docs that have results that say Fletcher32 is better than Adler32 because the prime modulus's "better bit mixing" benefits are outweighed by the reduced hash range as I speculated back when doing pysync;

http://minkirri.apana.org.au/~abo/projects/pysync/adler32.py

librsync uses a modified fletcher32 with no modulus, 2's compliment addition with a "block size" of 8 plus a prime offset of 31. Note the prime offset means the "s1 is is initialized to 1" part is addressed; any sequence of zero's is not zero. I dunno if this is actually better than just initializing s1=1, but it is not worse.

The use of a block size of 8 (instead of 16) means we can only overflow the s1 counter at most every 2^16 / (2^8 - 1 +31) = 229 bytes, which reduces the difference between 1's and 2's compliment there (which only differ on overflow... 1's compliment would +1 every overflow), though s2 would still overflow pretty quickly.

I think the modulus prime has been shown to not be worth it in practice. The only question is should we add a mod 2^16-1? Again, my gut feeling is it will not be worth it.

In summary, I think the existing rollsum is good enough... and probably better than adler32. I would have to see pretty good evidence to convince me otherwise.

**Comment by dbaarda** [comment]
For the record; 16bit 1's compliment addition is the same as 2's compliment addition mod 0xffff, that's where those bits of Fletcher come from. 

It's better than just 2's compliment for fletcher because it ensures all bits are treated equally... single bit errors in 2's compliment only affect the bits above them so low bit errors affect more of the sum than high-bit errors, but 1's compliment has an "end-around carry" that propogates high-bit error effects back to the lower end bits. Another way to say this is the 0xffff  modulus used is relatively prime to all 16 bit positions.

So adding a mod 0xffff might improve things a little bit, but since we are adding 8bit values to 16bit sums, even errors in the 8th bit will affect 8/16 of the s
